File: data/training_collector_v2.go
package main

import (
	"bufio"
	"crypto/md5"
	"flag"
	"fmt"
	"log"
	"math/rand"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"github.com/xitongsys/parquet-go/parquet"
	"github.com/xitongsys/parquet-go-source/local"
	"github.com/xitongsys/parquet-go/writer"
)

// SMT2File holds the parsed contents of an SMT2 file.
type SMT2File struct {
	Constants string
	Solution  string
}

// QuestionEntry is the structure we output in the Parquet file.
type QuestionEntry struct {
	Problem         string `parquet:"name=problem, type=BYTE_ARRAY, convertedtype=UTF8"`
	ExampleIndices  string `parquet:"name=example_indices, type=BYTE_ARRAY, convertedtype=UTF8"`
	Examples        string `parquet:"name=examples, type=BYTE_ARRAY, convertedtype=UTF8"`
	Question        string `parquet:"name=question, type=BYTE_ARRAY, convertedtype=UTF8"`
	AnswerIndex     int    `parquet:"name=answer_index, type=INT32"`
	AnswerConstants string `parquet:"name=answer_constants, type=BYTE_ARRAY, convertedtype=UTF8"`
	AnswerSolution  string `parquet:"name=answer_solution, type=BYTE_ARRAY, convertedtype=UTF8"`
	VariableMapping string `parquet:"name=variable_mapping, type=BYTE_ARRAY, convertedtype=UTF8"`
}

// IndexPair represents a combination (list of indices) and an answer index.
type IndexPair struct {
	Combo       []int
	AnswerIndex int
}

// parseSMT2File reads an SMT2 file and returns its constants and solution.
func parseSMT2File(path string) (SMT2File, error) {
	f, err := os.Open(path)
	if err != nil {
		return SMT2File{}, err
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	var constLines, solLines []string
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if strings.HasPrefix(line, "(declare-const") {
			constLines = append(constLines, line)
		} else if strings.HasPrefix(line, "(assert") {
			solLines = append(solLines, line)
		}
	}
	if err := scanner.Err(); err != nil {
		return SMT2File{}, err
	}
	return SMT2File{
		Constants: strings.Join(constLines, "\n"),
		Solution:  strings.Join(solLines, "\n"),
	}, nil
}

// groupSMT2Files walks the base directory and groups SMT2 file paths by problem name and index.
// Expected file name pattern: own.<problem>_<N>.smt2
func groupSMT2Files(baseDir string) (map[string]map[int]string, error) {
	problemsFiles := make(map[string]map[int]string)
	pattern := regexp.MustCompile(`^own\.(.+)_(\d+)\.smt2$`)
	err := filepath.Walk(baseDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			return nil
		}
		if filepath.Ext(info.Name()) == ".smt2" {
			matches := pattern.FindStringSubmatch(info.Name())
			if len(matches) == 3 {
				problem := matches[1]
				idx, err := strconv.Atoi(matches[2])
				if err != nil {
					return err
				}
				if _, ok := problemsFiles[problem]; !ok {
					problemsFiles[problem] = make(map[int]string)
				}
				problemsFiles[problem][idx] = path
			}
		}
		return nil
	})
	return problemsFiles, err
}

// extractVariableNames returns unique variable names found in constantsText.
func extractVariableNames(constantsText string) []string {
	varNamesSet := make(map[string]struct{})
	pattern := regexp.MustCompile(`\(declare-const\s+(\S+)`)
	matches := pattern.FindAllStringSubmatch(constantsText, -1)
	for _, match := range matches {
		if len(match) > 1 {
			varNamesSet[match[1]] = struct{}{}
		}
	}
	var varNames []string
	for k := range varNamesSet {
		varNames = append(varNames, k)
	}
	sort.Strings(varNames)
	return varNames
}

// renameVariablesInTexts renames variables in each text (and in the constants) based on a random mapping.
func renameVariablesInTexts(texts []string, constantsText string, seed int64) ([]string, string, map[string]string) {
	mapping := make(map[string]string)
	if constantsText != "" {
		varNames := extractVariableNames(constantsText)
		rng := rand.New(rand.NewSource(seed))
		prefixes := []string{"in", "var", "tmp", "aux", "res", "out", "arg", "param", "local", "global"}
		// Also add single letters a-z.
		for ch := 'a'; ch <= 'z'; ch++ {
			prefixes = append(prefixes, string(ch))
		}
		newNames := make([]string, len(varNames))
		for i := range varNames {
			prefix := prefixes[rng.Intn(len(prefixes))]
			newNames[i] = fmt.Sprintf("%s_%d", prefix, i+1)
		}
		for i, v := range varNames {
			mapping[v] = newNames[i]
		}
	}

	var replaceFunc func(string) string
	if len(mapping) > 0 {
		var keys []string
		for k := range mapping {
			keys = append(keys, regexp.QuoteMeta(k))
		}
		pattern := regexp.MustCompile(`\b(` + strings.Join(keys, "|") + `)\b`)
		replaceFunc = func(text string) string {
			if text == "" {
				return text
			}
			return pattern.ReplaceAllStringFunc(text, func(match string) string {
				return mapping[match]
			})
		}
	} else {
		replaceFunc = func(text string) string {
			return text
		}
	}

	renamedTexts := make([]string, len(texts))
	for i, t := range texts {
		renamedTexts[i] = replaceFunc(t)
	}
	renamedConstants := ""
	if constantsText != "" {
		renamedConstants = replaceFunc(constantsText)
	}
	return renamedTexts, renamedConstants, mapping
}

// makePrefix constructs the prompt text given the examples and answer index.
func makePrefix(model, examples string, NQuestion int, instruct bool) string {
	if !strings.HasPrefix(model, "Qwen") {
		log.Fatal("Only Qwen models are supported.")
	}
	if instruct {
		return fmt.Sprintf(`<|im_start|>system
    You are a helpful assistant.
    You first think about the reasoning process in your mind and then provide the user with the answer.
    <|im_end|>
    <|im_start|>user
    Your role is to take a known pattern of symbolic constraints that represent the longest execution path of a program 
    and generalize it for any given input size N.
    When you receive an input value N,
    you must generate a canonical SMT-LIB constraint string that adheres to the following rules:
    (assert (op (op (op var_1 var_2)) (op (op var_3 var_4)) (op (op var_5 var_6)) (op var_7 var_8)))
    where op is a logical operator (e.g., 'and', 'or', 'not') and var_i are variables or constants.
    All per-variable constraints must be combined using a top-level (assert (and ...)) clause.
    The output must be in exact, canonical SMT-LIB format without extra commentary in the constraint string.
    Show your work in <think> </think> tags. And return the final SMT-LIB constraint string in <answer> </answer> tags.
    For example: <answer>(assert (and  ( >=  in0 97)  ( <=  in0 122)))</answer>.
    Here are the known constraints:
    %s
    What is the constraint for N=%d?
    <|im_end|>
    <|im_start|>assistant
    Let me solve this step by step.
    <think>`, examples, NQuestion)
	}
	return fmt.Sprintf(`A conversation between User and Assistant. The user asks a question, and the Assistant solves it.
The assistant first thinks about the reasoning process in the mind and then provides the user with the answer.
User: Your role is to take a known pattern of symbolic constraints that represent the longest execution path of a program 
and generalize it for any given input size N.
When you receive an input value N,
you must generate a canonical SMT-LIB constraint string that adheres to the following rules:
(assert (op (op (op var_1 var_2)) (op (op var_3 var_4)) (op (op var_5 var_6)) (op var_7 var_8)))
where op is a logical operator (e.g., 'and', 'or', 'not') and var_i are variables or constants.
All per-variable constraints must be combined using a top-level (assert (and ...)) clause.
The output must be in exact, canonical SMT-LIB format without extra commentary in the constraint string.
Show your work in <think> </think> tags. And return the final SMT-LIB constraint string in <answer> </answer> tags.
For example: <answer>(assert (and  ( >=  in0 97)  ( <=  in0 122)))</answer>.
Here are the known constraints:
%s
What is the constraint for N=%d?
Assistant: Let me solve this step by step.
<think>`, examples, NQuestion)
}

// generateIndexPairs returns all (combo, answerIndex) pairs.
func generateIndexPairs(max_n, min_examples, max_examples, difference int) []IndexPair {
	var pairs []IndexPair
	candidateIndices := make([]int, max_n)
	for i := 0; i < max_n; i++ {
		candidateIndices[i] = i + 1
	}
	for k := min_examples; k <= max_examples; k++ {
		combs := combinations(candidateIndices, k)
		for _, combo := range combs {
			lastExample := combo[len(combo)-1]
			for d := 1; d <= difference; d++ {
				answerIndex := lastExample + d
				if answerIndex <= max_n {
					pairs = append(pairs, IndexPair{Combo: combo, AnswerIndex: answerIndex})
				}
			}
		}
	}
	return pairs
}

// combinations returns all combinations of length k from arr.
func combinations(arr []int, k int) [][]int {
	var res [][]int
	var comb func(start int, curr []int)
	comb = func(start int, curr []int) {
		if len(curr) == k {
			tmp := make([]int, k)
			copy(tmp, curr)
			res = append(res, tmp)
			return
		}
		for i := start; i < len(arr); i++ {
			comb(i+1, append(curr, arr[i]))
		}
	}
	comb(0, []int{})
	return res
}

// md5Seed computes an int64 seed from the MD5 hash of input.
func md5Seed(input string) int64 {
	h := md5.New()
	h.Write([]byte(input))
	sum := h.Sum(nil)
	var seed int64 = 0
	// Use the first 8 bytes.
	for i := 0; i < 8 && i < len(sum); i++ {
		seed = (seed << 8) | int64(sum[i])
	}
	return seed
}

// generateQuestionEntriesForProblem creates question entries for one problem.
func generateQuestionEntriesForProblem(files map[int]string, problem string, globalSeed int64, instruct bool, min_examples, max_examples, difference, max_n int) ([]QuestionEntry, error) {
	parsedFiles := make(map[int]SMT2File)
	for idx, path := range files {
		fileContent, err := parseSMT2File(path)
		if err != nil {
			return nil, err
		}
		parsedFiles[idx] = fileContent
	}
	var entries []QuestionEntry
	pairs := generateIndexPairs(max_n, min_examples, max_examples, difference)
	for _, pair := range pairs {
		comboStr := fmt.Sprintf("%v", pair.Combo)
		seedInput := fmt.Sprintf("%s_%s_%d_%d", problem, comboStr, pair.AnswerIndex, globalSeed)
		// Removed localSeed since it's unused:
		_ = md5Seed(seedInput)

		answerParsed, ok := parsedFiles[pair.AnswerIndex]
		if !ok {
			answerParsed = SMT2File{Constants: "", Solution: ""}
		}

		type Example struct {
			Index    int
			Solution string
		}
		var examples []Example
		for _, i := range pair.Combo {
			f, ok := parsedFiles[i]
			if !ok {
				f = SMT2File{Constants: "", Solution: ""}
			}
			examples = append(examples, Example{Index: i, Solution: f.Solution})
		}

		var allTexts []string
		for _, ex := range examples {
			allTexts = append(allTexts, ex.Solution)
		}
		allTexts = append(allTexts, answerParsed.Solution)

		renameSeedInput := fmt.Sprintf("%s_%s_%d_rename_%d", problem, comboStr, pair.AnswerIndex, globalSeed)
		renameSeed := md5Seed(renameSeedInput)

		renamedTexts, renamedAnswerConstants, mapping := renameVariablesInTexts(allTexts, answerParsed.Constants, renameSeed)

		// Build examples string and also record the indices.
		var renamedExamples []string
		var exampleIndices []string
		for i, ex := range examples {
			renamedExamples = append(renamedExamples, fmt.Sprintf("N=%d: %s", ex.Index, renamedTexts[i]))
			exampleIndices = append(exampleIndices, strconv.Itoa(ex.Index))
		}
		examplesStr := strings.Join(renamedExamples, "\n")
		questionText := makePrefix("Qwen", examplesStr, pair.AnswerIndex, instruct)

		// Prepare variable mapping as a simple comma-separated key:value string.
		var mappingParts []string
		for k, v := range mapping {
			mappingParts = append(mappingParts, fmt.Sprintf("%s:%s", k, v))
		}
		mappingStr := strings.Join(mappingParts, ",")

		entry := QuestionEntry{
			Problem:         problem,
			ExampleIndices:  strings.Join(exampleIndices, ","),
			Examples:        examplesStr,
			Question:        questionText,
			AnswerIndex:     pair.AnswerIndex,
			AnswerConstants: renamedAnswerConstants,
			AnswerSolution:  renamedTexts[len(renamedTexts)-1],
			VariableMapping: mappingStr,
		}
		entries = append(entries, entry)
	}
	return entries, nil
}

// collectSMT2DataFrame groups files by problem and generates all question entries.
func collectSMT2DataFrame(baseDir string, globalSeed int64, instruct bool) ([]QuestionEntry, error) {
	problemsFiles, err := groupSMT2Files(baseDir)
	if err != nil {
		return nil, err
	}
	var allEntries []QuestionEntry
	for problem, files := range problemsFiles {
		fmt.Printf("Processing problem: %s\n", problem)
		entries, err := generateQuestionEntriesForProblem(files, problem, globalSeed, instruct, 3, 5, 3, 30)
		if err != nil {
			return nil, err
		}
		allEntries = append(allEntries, entries...)
	}
	fmt.Printf("Collected %d question instances.\n", len(allEntries))
	return allEntries, nil
}

func main() {
	baseDir := flag.String("base_dir", "", "Base directory to search for SMT2 files")
	output := flag.String("output", "output.parquet", "Output Parquet file")
	seed := flag.Int64("seed", 69420, "Random seed for reproducibility")
	instruct := flag.Bool("instruct", false, "Whether to follow Qwen instruction format")
	flag.Parse()

	if *baseDir == "" {
		log.Fatal("base_dir is required")
	}

	entries, err := collectSMT2DataFrame(*baseDir, *seed, *instruct)
	if err != nil {
		log.Fatalf("Error collecting data: %v", err)
	}

	// Write the results to a Parquet file.
	fw, err := local.NewLocalFileWriter(*output)
	if err != nil {
		log.Fatalf("Can't create local file writer: %v", err)
	}
	pw, err := writer.NewParquetWriter(fw, new(QuestionEntry), 4)
	if err != nil {
		log.Fatalf("Can't create parquet writer: %v", err)
	}
	pw.RowGroupSize = 128 * 1024 * 1024
	pw.CompressionType = parquet.CompressionCodec_SNAPPY

	for _, entry := range entries {
		if err = pw.Write(entry); err != nil {
			log.Println("Write error", err)
		}
	}
	if err = pw.WriteStop(); err != nil {
		log.Fatalf("WriteStop error: %v", err)
	}
	fw.Close()
	fmt.Println("Parquet file written successfully.")
}

File: data/training_collector_v2.py
import argparse
import hashlib
import random
import re
from pathlib import Path
from tqdm import tqdm

import pandas as pd


def parse_smt2_file(filepath: Path) -> dict[str, str]:
    """
    Parse an SMT2 file to extract constants and solution.
    Lines starting with "(declare-const" are treated as constants.
    Lines starting with "(assert" are treated as part of the solution.
    """
    constants_lines = []
    solution_lines = []
    with filepath.open("r") as f:
        for line in f:
            stripped = line.strip()
            if stripped.startswith("(declare-const"):
                constants_lines.append(stripped)
            elif stripped.startswith("(assert"):
                solution_lines.append(stripped)
    return {"constants": "\n".join(constants_lines), "solution": "\n".join(solution_lines)}


def group_smt2_files(base_dir: str) -> dict[str, dict[int, Path]]:
    """
    Walk through the base directory and group SMT2 files by problem name and index.
    Files are expected to be named as: own.<problem>_<N>.smt2.
    """
    base_path = Path(base_dir)
    pattern = re.compile(r"^own\.(?P<problem>.+)_(?P<index>\d+)\.smt2$")
    problems_files: dict[str, dict[int, Path]] = {}

    for filepath in base_path.rglob("*.smt2"):
        match = pattern.match(filepath.name)
        if match:
            problem = match.group("problem")
            index = int(match.group("index"))
            if problem not in problems_files:
                problems_files[problem] = {}
            problems_files[problem][index] = filepath

    return problems_files


def extract_variable_names(constants_text: str) -> set[str]:
    """
    Extract variable names from constants text using the pattern in declare-const lines.
    """
    if not constants_text:
        return set()
    return set(re.findall(r"\(declare-const\s+(\S+)", constants_text))


def rename_variables_in_texts(texts: list[str], constants_text: str, seed: int):
    """
    Rename variable names in a list of texts (examples + answer solution)
    based on variable names extracted from constants_text.

    Returns a tuple: (renamed_texts, renamed_constants_text, mapping)
    """
    mapping = {}
    if constants_text:
        var_names = list(extract_variable_names(constants_text))
        rng = random.Random(seed)
        # Use a diverse set of candidate prefixes.
        prefixes = ["in", "var", "tmp", "aux", "res", "out", "arg", "param", "local", "global"]
        prefixes.extend(list("abcdefghijklmnopqrstuvwxyz"))
        new_names = [f"{rng.choice(prefixes)}_{i+1}" for i in range(len(var_names))]
        mapping = dict(zip(var_names, new_names))

    if mapping:
        pattern = re.compile(r"\b(" + "|".join(re.escape(key) for key in mapping.keys()) + r")\b")

        def replace_vars(text: str) -> str:
            if not text:
                return text
            return pattern.sub(lambda m: mapping[m.group(0)], text)
    else:
        def replace_vars(text: str) -> str:
            return text

    renamed_texts = [replace_vars(t) for t in texts]
    renamed_constants_text = replace_vars(constants_text) if constants_text else None

    return renamed_texts, renamed_constants_text, mapping


def make_prefix(model: str, examples: str, N_question: int, instruct: bool) -> str:
    """
    Construct the prompt that wraps example constraints and asks for an answer for N_question.
    """
    if not model.startswith("Qwen"):
        raise NotImplementedError("Only Qwen models are supported.")

    if instruct:
        return f"""<|im_start|>system
    You are a helpful assistant.
    You first think about the reasoning process in your mind and then provide the user with the answer.
    <|im_end|>
    <|im_start|>user
    Your role is to take a known pattern of symbolic constraints that represent the longest execution path of a program 
    and generalize it for any given input size N.
    When you receive an input value N,
    you must generate a canonical SMT-LIB constraint string that adheres to the following rules:
    (assert (op (op (op var_1 var_2)) (op (op var_3 var_4)) (op (op var_5 var_6)) (op var_7 var_8)))
    where op is a logical operator (e.g., 'and', 'or', 'not') and var_i are variables or constants.
    All per-variable constraints must be combined using a top-level (assert (and ...)) clause.
    The output must be in exact, canonical SMT-LIB format without extra commentary in the constraint string.
    Show your work in <think> </think> tags. And return the final SMT-LIB constraint string in <answer> </answer> tags.
    For example: <answer>(assert (and  ( >=  in0 97)  ( <=  in0 122)))</answer>.
    Here are the known constraints:
    {examples}
    What is the constraint for N={N_question}?
    <|im_end|>
    <|im_start|>assistant
    Let me solve this step by step.
    <think>"""

    # "chat" style format if not instruct
    return f"""A conversation between User and Assistant. The user asks a question, and the Assistant solves it.
User: Your role is to take a known pattern of symbolic constraints that represent the longest execution path of a program 
and generalize it for any given input size N.
When you receive an input value N,
you must generate a canonical SMT-LIB constraint string that adheres to the following rules:
(assert (op (op (op var_1 var_2)) (op (op var_3 var_4)) (op (op var_5 var_6)) (op var_7 var_8)))
where op is a logical operator (e.g., 'and', 'or', 'not') and var_i are variables or constants.
All per-variable constraints must be combined using a top-level (assert (and ...)) clause.
The output must be in exact, canonical SMT-LIB format without extra commentary in the constraint string.
Show your work in <think> </think> tags. And return the final SMT-LIB constraint string in <answer> </answer> tags.
For example: <answer>(assert (and  ( >=  in0 97)  ( <=  in0 122)))</answer>.
Here are the known constraints:
{examples}
What is the constraint for N={N_question}?
Assistant: Let me solve this step by step.
<think>"""


def generate_consecutive_index_pairs(max_n: int, min_examples: int, max_examples: int, difference: int):
    """
    Yields only *consecutive* combos of size k in [min_examples..max_examples].
    For each combo, yields up to 'difference' possible answer_indices = last_example + d,
    so long as (answer_index <= max_n).

    For example, if k=3 and difference=3, (1,2,3) can pair with answer=4,5,6 if they are <= max_n.
    Then (2,3,4) pairs with answer=5,6,7, etc.
    """
    for k in range(min_examples, max_examples + 1):
        # consecutive runs of length k
        for start in range(1, max_n - k + 2):
            combo = tuple(range(start, start + k))
            last_example = combo[-1]
            for d in range(1, difference + 1):
                answer_index = last_example + d
                if answer_index <= max_n:
                    yield combo, answer_index


def generate_question_entries_for_problem(
    files_dict: dict[int, Path],
    problem: str,
    global_seed: int,
    instruct: bool,
    min_examples: int = 1,
    max_examples: int = 10,
    difference: int = 3,
    max_n: int = 30
) -> list[dict]:
    """
    For a single problem, generate question instances by:
      1) Generating *consecutive* index combos (rather than all combinations).
      2) Caching parsed file contents.
      3) Renaming variables to avoid collisions.
      4) Building question + answer entries with metadata.
    """
    # Pre-cache parsed file data to avoid re-reading the same file multiple times.
    parsed_files = {i: parse_smt2_file(path) for i, path in files_dict.items()}

    question_entries = []

    for combo, answer_index in generate_consecutive_index_pairs(max_n, min_examples, max_examples, difference):
        # Deterministic seed for per-(combo, answer_index)
        seed_input = f"{problem}_{combo}_{answer_index}_{global_seed}"
        local_seed = int(hashlib.md5(seed_input.encode()).hexdigest(), 16) % (2**32)

        # Grab the answer data, if it exists
        answer_parsed = parsed_files.get(answer_index, {"constants": None, "solution": None})

        # Get each example from cache
        examples = []
        for i in combo:
            parsed = parsed_files.get(i, {"constants": None, "solution": None})
            examples.append({"index": i, "solution": parsed["solution"]})

        # Combine example solutions + the answer solution for consistent variable renaming
        all_texts = [ex["solution"] for ex in examples] + [answer_parsed["solution"]]
        rename_seed_input = f"{problem}_{combo}_{answer_index}_rename_{global_seed}"
        rename_seed = int(hashlib.md5(rename_seed_input.encode()).hexdigest(), 16) % (2**32)

        renamed_texts, renamed_answer_constants, mapping = rename_variables_in_texts(
            all_texts,
            answer_parsed["constants"],
            rename_seed
        )

        # Separate the renamed solutions back out
        renamed_examples = [
            {"index": ex["index"], "solution": renamed_texts[idx]}
            for idx, ex in enumerate(examples)
        ]
        renamed_answer_solution = renamed_texts[-1]

        # Build the multi-example "Here are the known constraints" string
        examples_str = "\n".join(
            f"N={ex['index']}: {ex['solution']}" for ex in renamed_examples
        )
        # Create the Q&A prompt
        question_text = make_prefix("Qwen", examples_str, answer_index, instruct)

        entry = {
            "problem": problem,
            "example_indices": list(combo),
            "examples": renamed_examples,
            "question": question_text,
            "answer_index": answer_index,
            "answer_constants": renamed_answer_constants,
            "answer_solution": renamed_answer_solution,
            "variable_mapping": mapping,
        }
        question_entries.append(entry)

    return question_entries


def collect_smt2_dataframe(base_dir: str, global_seed: int, instruct: bool) -> pd.DataFrame:
    """
    Collect and organize SMT2 data from the specified directory into a DataFrame.
    """
    problems_files = group_smt2_files(base_dir)
    all_entries = []
    for problem, files_dict in tqdm(problems_files.items()):
        # Here you can adjust parameters as needed:
        entries = generate_question_entries_for_problem(
            files_dict,
            problem,
            global_seed,
            instruct,
            min_examples=1,     # e.g. only run length 3
            max_examples=5,     # you could allow up to 5 if you want (3..5)
            difference=5,
            max_n=30
        )
        all_entries.extend(entries)
    print(f"Collected {len(all_entries)} question instances.")
    return pd.DataFrame(all_entries)


def parse_arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Collect and organize SMT2 file data into a DataFrame and save as a Parquet file."
    )
    parser.add_argument("base_dir", type=str, help="Base directory to search for SMT2 files.")
    parser.add_argument("output", type=str, help="Output Parquet file.")
    parser.add_argument("--seed", type=int, default=69_420, help="Random seed for reproducibility.")
    parser.add_argument("--instruct", action="store_true", help="Whether to follow Qwen instruction format.")
    return parser.parse_args()


def main() -> None:
    args = parse_arguments()
    df = collect_smt2_dataframe(args.base_dir, args.seed, args.instruct)
    df.to_parquet(args.output)


if __name__ == "__main__":
    main()
